### PR TITLE
Consistent Address Types

### LINKED ISSUE #3091: Implement solution to EVM working with only one type of address (evm_address)
1. Normally created address, without an association: gets processed and accessed through its long-zero address
2. Address created via the create2 operation: 
    1. When called through HAPI, it doesn’t matter which address (mirror/create2) we refer it with.
    2. When called through the EVM, only the create2 address will have to be used.

Initial proposal: https://limechain.notion.site/Consistent-EVM-addresses-084b2c8491b146379666f8540ad149b3

### PR DESCRIPTION
**Pull request by georgiyazovaliiski**
**Description**:
On a weekly sync there it was discussed that there should be consistency when referring to addresses from inside the EVM.

**Where can addresses be seen?**

1. Contract storage (fields of the contract)
2. Function parameters
3. msg.sender and all other predefined contract properties

A contract can be called **both** from **HAPI** with a long-zero address and through the **EVM** directly (using create2 address).

In order to avoid confusion, the **EVM** in itself will act as if long-zero addresses, which have an associated alias, do not exist.

Addresses get accessed in the Hedera services project through the get() method in AbstractLedgerWorldUpdater.

There we need to add logic which checks whether the address we are checking is a “**mirrorAddress**” and there IS an **associated alias** for it. If this is true, the address is ignored and null is returned.

*TL;DR version*:

1. Normally created address, without an association: gets processed and accessed through its long-zero address
2. Address created via the create2 operation: 
    1. When called through HAPI, it doesn’t matter which address (mirror/create2) we refer it with.
    2. When called through the EVM, only the create2 address will have to be used.

Fixes #3091

**Notes**
There are some lines which are not covered. They should get covered by another PR. Everything that uses logic written solely for the purpose of this PR has been covered.

**Checklist**

- [ ] Adapt for Hedera aliased accounts (0.0.key)
- [x] Unit tests
- [x] Current E2E Tests adapted
- [x] New E2E Scenario written (assertions need to be adjusted but the scenario is there)

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/contracts/operation/HederaExtCodeHashOperation.java [hunk 1]
```diff
@@ -32,6 +33,7 @@
 import org.hyperledger.besu.evm.internal.FixedStack;
 import org.hyperledger.besu.evm.internal.Words;
 import org.hyperledger.besu.evm.operation.ExtCodeHashOperation;
+import org.hyperledger.besu.evm.operation.Operation;
```
Comment (IvanKavaldzhiev) [inline hedera-node/src/main/java/com/hedera/services/contracts/operation/HederaExtCodeHashOperation.java:36]: Probably these imports are redundant?

File: hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java [hunk 2]
```diff
@@ -147,12 +147,24 @@ public EvmAccount createAccount(final Address addressOrAlias, final long nonce,
...
+
 		final var extantMutable = this.updatedAccounts.get(address);
 		if (extantMutable != null) {
 			return extantMutable;
 		} else {
-			return this.deletedAccounts.contains(address) ? null : this.world.get(address);
+			if (this.deletedAccounts.contains(address)) {
+				return null;
+			}
+			if (this.world.getClass() == HederaWorldState.class) {
```
Comment (tinker-michaelj) [inline hedera-node/src/main/java/com/hedera/services/store/contracts/AbstractLedgerWorldUpdater.java:164]: Can we just do 
```suggestion
			if (this.world instanceof HederaWorldState) {
```

File: test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java [hunk 3]
```diff
@@ -193,9 +200,61 @@ public List<HapiApiSpec> getSpecsInSuite() {
...
+										getNestedContractAddress(WHITELISTER, spec))
+										.payingWith(DEFAULT_PAYER)
+										.via(mirrorWhitelistCheckTxn),
+								contractCall(CREATOR, IS_WHITELISTED_ABI,
+										getNestedContractAddress(WHITELISTER, spec))
+										.payingWith(DEFAULT_PAYER)
+										.via(evmWhitelistCheckTxn)
+						))
+				).then(
+//						getTxnRecord(mirrorWhitelistCheckTxn).hasPriority(recordWith().contractCallResult(resultWith().contractCallResult(simpleBytes(Bytes.of(1))))).logged(),
```
Comment (tinker-michaelj) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/contract/hapi/ContractCallSuite.java:253]: Were these meant to be commented out?

**Review by tinker-michaelj — APPROVED** [review]
LGTM, very nice work! I think there's one item left re: the address passed to a `selfdestruct`, but can be done in a separate PR.
